### ISSUE TITLE
return address 

### ISSUE THREAD

**Issue by DieracDelta** [draperlaboratory/cbat_tools#213]
I've got [main_1](https://gist.github.com/98ea38abde6afa9f4b346db123e2ccc5), [main_2](https://gist.github.com/733c72e06888f8dc637cf6096ad2c294), an [invocation](https://gist.github.com/bc22ed148646846cd57717faeefa46b5), the main_1 [cfg](https://gist.github.com/cf91dfe7ff8d0a544035a7957e156a4f), and the main_2 [cfg](https://gist.github.com/4b7e8f3ac0ad6a7563533bff758fed4a). This comes back SAT with [this](https://gist.github.com/e0f48ba8f06042a69ca2f17284771a3c) countermodel and it looks like the return address being pushed onto the stack right before the function call (tid %00000026) is the culprit. I believe this is because CBAT chooses a value for RBX that is on the stack and includes part of this return address. mem[rbx] is then read into a temporary variable. And then that temporary variable is then moved into RAX, which is our refuted goal. I did confirm that the tid has a spec (_start has `spec_arg_terms`), and thus stack pointer is incremented. However, if one removes the offset label and instruction in `main_2` that cause the pushed return address to be different, the programs are unsat. Does this make sense?

This feels like something that can be corrected by adding in an assertion afterwards that the stack memory below RSP after a function is called is equal between programs. However, maybe this is this too strong of an assumption?

**Comment by DieracDelta** [comment]
I'm pasting in the gist metadata. `main_1`'s code :
```
.global _start

.text

safe_hasher:

push %rbx
#callq *0x30(%rbx)
callq _start
add (%rbx),%rax
pop %rbx
retq

main:
    ret

_start:
    call    main
```
`main_2`'s code:
```
.global _start

.text

this_creates_an_offset:
push %rbx

safe_hasher:

push %rbx
#callq *0x30(%rbx)
callq _start
add (%rbx),%rax
pop %rbx
retq

main:
    ret

_start:
    call    main
```
`main_1`'s cfg:
```
digraph {
node[shape=box];

subgraph "cluster__start" {
  label="_start"
  
  "\%0000001c"[label="\%0000001c
0000002f: RSP := RSP - 8\l00000032: mem := mem with [RSP, el]:u64 <- 0x4000E5\l00000035: call @main with noreturn\l"]
}
"\%0000001c" -> "\%0000002b";

subgraph "cluster___libc_start_main" {
  label="__libc_start_main"
  
  "\%0000002b"[label="\%0000002b
0000003d: #2 := mem[RSP, el]:u64\l00000040: RSP := RSP + 8\l00000044: call #2 with noreturn\l"]
}

subgraph "cluster_safe_hasher" {
  label="safe_hasher"
  
  "\%00000008"[label="\%00000008
0000000f: #1 := RBX\l00000013: RSP := RSP - 8\l00000016: mem := mem with [RSP, el]:u64 <- #1\l00000020: RSP := RSP - 8\l00000023: mem := mem with [RSP, el]:u64 <- 0x4000DA\l00000026: call @_start with return %00000046\l"]
  "\%00000046"[label="\%00000046
00000053: #3 := RAX\l00000056: #4 := mem[RBX, el]:u64\l00000059: RAX := RAX + #4\l0000005c: CF := RAX < #3\l0000005f: OF := high:1[#3] = high:1[#4] & (high:1[#3] | high:1[RAX]) & ~(\lhigh:1[#3] & high:1[RAX])\l00000062: AF := 0x10 = (0x10 & (RAX ^ #3 ^ #4))\l00000065: PF := ~low:1[let $1 = RAX >> 4 ^ RAX in let $2 = $1 >> 2 ^ $1 in\l$2 >> 1 ^ $2]\l00000068: SF := high:1[RAX]\l0000006b: ZF := 0 = RAX\l00000073: RBX := mem[RSP, el]:u64\l00000076: RSP := RSP + 8\l0000007f: #6 := mem[RSP, el]:u64\l00000082: RSP := RSP + 8\l00000086: call #6 with noreturn\l"]
  "\%00000008" -> "\%00000046"
}
"\%00000008" -> "\%0000001c";
}
```
`main_2`'s cfg:
```
digraph {
node[shape=box];

subgraph "cluster__start" {
  label="_start"
  
  "\%0000002b"[label="\%0000002b
0000003e: RSP := RSP - 8\l00000041: mem := mem with [RSP, el]:u64 <- 0x4000E6\l00000044: call @main with noreturn\l"]
}
"\%0000002b" -> "\%0000003a";

subgraph "cluster___libc_start_main" {
  label="__libc_start_main"
  
  "\%0000003a"[label="\%0000003a
0000004c: #3 := mem[RSP, el]:u64\l0000004f: RSP := RSP + 8\l00000053: call #3 with noreturn\l"]
}

subgraph "cluster_safe_hasher" {
  label="safe_hasher"
  
  "\%00000018"[label="\%00000018
0000001f: #2 := RBX\l00000022: RSP := RSP - 8\l00000025: mem := mem with [RSP, el]:u64 <- #2\l0000002f: RSP := RSP - 8\l00000032: mem := mem with [RSP, el]:u64 <- 0x4000DB\l00000035: call @_start with return %00000055\l"]
  "\%00000055"[label="\%00000055
00000062: #4 := RAX\l00000065: #5 := mem[RBX, el]:u64\l00000068: RAX := RAX + #5\l0000006b: CF := RAX < #4\l0000006e: OF := high:1[#4] = high:1[#5] & (high:1[#4] | high:1[RAX]) & ~(\lhigh:1[#4] & high:1[RAX])\l00000071: AF := 0x10 = (0x10 & (RAX ^ #4 ^ #5))\l00000074: PF := ~low:1[let $1 = RAX >> 4 ^ RAX in let $2 = $1 >> 2 ^ $1 in\l$2 >> 1 ^ $2]\l00000077: SF := high:1[RAX]\l0000007a: ZF := 0 = RAX\l00000082: RBX := mem[RSP, el]:u64\l00000085: RSP := RSP + 8\l0000008e: #7 := mem[RSP, el]:u64\l00000091: RSP := RSP + 8\l00000095: call #7 with noreturn\l"]
  "\%00000018" -> "\%00000055"
}
"\%00000018" -> "\%0000002b";

subgraph "cluster_this_creates_an_offset" {
  label="this_creates_an_offset"
  
  "\%00000008"[label="\%00000008
0000000f: #1 := RBX\l00000013: RSP := RSP - 8\l00000016: mem := mem with [RSP, el]:u64 <- #1\l0000009b: call @safe_hasher with noreturn\l"]
}
"\%00000008" -> "\%00000018";
}
```
Note that the only difference between `main_1` and `main_2` is the offset symbol and instruction in the second binary such that the return address pushed on in the bir differs between binaries.

The bap invocation:
```
gcc  -nostdlib main_1.S -o main_1
gcc  -nostdlib main_2.S -o main_2
bap --pass=save-project --no-byteweight --save-project-filename=main_1.bpj main_1
bap --pass=save-project --no-byteweight --save-project-filename=main_2.bpj main_2

BAP_DEBUG=1 bap /bin/echo --pass=wp --wp-compare=true --wp-file1=main_1.bpj --wp-file2=main_2.bpj --wp-function=safe_hasher --wp-inline=false --wp-num-unroll=0 --no-byteweight --wp-use-fun-input-regs=false --wp-mem-offset=false --wp-compare-post-reg-values=R12,R13,R14,R15,RBX,RSP,RBP,RAX --wp-show=bir,refuted-goals,paths
```
And finally, the resulting model:
```
Model:
	ZF  |->  0x0
	SF  |->  0x0
	RSP |->  0x0000000040000000
	RBX |->  0x000000003fffffed
	RAX |->  0x0000000000000000
	PF  |->  0x0
	OF  |->  0x0
	CF  |->  0x0
	AF  |->  0x0
	mem_orig |-> [
		else |-> 0x00]
	mem_mod |-> [
		#x000000003fffffed |-> 0x00 ;
		#x000000003fffffee |-> 0x00 ;
		#x000000003fffffef |-> 0x00 ;
		#x000000003ffffff0 |-> 0x00 ;
		#x000000003ffffff1 |-> 0x00 ;
		#x000000003ffffff2 |-> 0x00 ;
		#x000000003ffffff3 |-> 0x00 ;
		#x000000003ffffff4 |-> 0x00 ;
		#x000000003ffffff8 |-> 0x00 ;
		#x000000003ffffff9 |-> 0x00 ;
		#x000000003ffffffa |-> 0x00 ;
		#x000000003ffffffb |-> 0x00 ;
		#x000000003ffffffc |-> 0x00 ;
		#x000000003ffffffd |-> 0x00 ;
		#x000000003ffffffe |-> 0x00 ;
		#x000000003fffffff |-> 0x00 ;
		#x0000000040000000 |-> 0x00 ;
		#x0000000040000001 |-> 0x00 ;
		#x0000000040000002 |-> 0x00 ;
		#x0000000040000003 |-> 0x00 ;
		#x0000000040000004 |-> 0x00 ;
		#x0000000040000005 |-> 0x00 ;
		#x0000000040000006 |-> 0x00 ;
		#x0000000040000007 |-> 0x00 ;
		else |-> 0xda]




Refuted goals:
RAX0 = RAX09:
	Concrete values: = #x00004000da000000 #x00004000db000000
	Z3 Expression: = #x00004000da000000 #x00004000db000000
	Path:
```

**Comment by DieracDelta** [comment]
Initial state:

```
| address (8 bytes) | value |        |
|        0x40000000 |       | <- RSP |
|        0x3ffffff8 |       |        |
|        0x3ffffff0 |       |        |
|        0x3fffffe8 |       |        |
```


Then (see cfg), `RSP` is decremented:

```
RSP := RSP - 8

| address (8 bytes) | value |         |
|        0x40000000 |       |         |
|        0x3ffffff8 |       | <- RSP  |
|        0x3ffffff0 |       |         |
|        0x3fffffe8 |       |         |
```

Then, `RBX` is pushed onto the stack and decremented:

```
mem := mem with [RSP] <- \#1
RSP := RSP - 8

| address (8 bytes) | value    |        |
|        0x40000000 |          |        |
|        0x3ffffff8 | init_RBX |        |
|        0x3ffffff0 |          | <- RSP |
|        0x3fffffe8 |          |        |
```

Then the return address is pushed onto the stack:

```
mem := mem with [RSP] <- 0x4000DA/0x4000DB (DIFFERENT BETWEEN PROGRAMS)
```

```

| address (8 bytes) | value             |        |
|        0x40000000 |                   |        |
|        0x3ffffff8 | init_RBX          |        |
|        0x3ffffff0 | 0x4000DA/0x4000DB | <- RSP |
|        0x3fffffe8 |                   |        |

```

Then, the function is called:

````
call @_start with return %00000046
````

````
| address (8 bytes) | value             |        |
|        0x40000000 |                   |        |
|        0x3ffffff8 | init_RBX          | <- RSP |
|        0x3ffffff0 | 0x4000DA/0x4000DB |        |
|        0x3fffffe8 |                   |        |
````

Then:

```
\#4 = mem[RBX]
```

From the countermodel, `RBX` is `0x3fffffed`, so we read in some of the bytes from
`0x3ffffff0` (which is differing between binaries as that is the address of the
pushed return address)

```
RAX := RAX + \#4
```

RAX is not touched for the rest of the program and thus ends up in the
countermodel.

A more detailed model of that last step:

```
| address (8 bytes) |             value |        |
|        0x40000000 |                   |        |
|        0x3ffffff8 |          init_RBX | <- RSP |
|        0x3ffffff7 | 0x4000DA/0x4000DB |        |
|        0x3ffffff6 |                00 |        |
|        0x3ffffff5 |                00 |        |
|        0x3ffffff4 |                00 | <----+ |
|        0x3ffffff3 |                00 |      | |
|        0x3ffffff2 |                40 |      | |
|        0x3ffffff1 |                00 |      R |
|        0x3ffffff0 |             DA/DA |      B |
|        0x3fffffef |                   |      X |
|        0x3fffffee |                   |      | |
|        0x3fffffed |                   | <----+ |
|        0x3fffffec |                   |        |
|        0x3fffffeb |                   |        |
|        0x3fffffea |                   |        |
|        0x3fffffe9 |                   |        |
|        0x3fffffe8 |                   |        |
|                   |                   |        |
```

**Comment by fortunac** [comment]
I took a quick look at this, and it does look like having different values in RAX is the correct behavior if the initial value of RBX is pointing to uninitialized places on the stack. Saying that the memory on the stack below RSP is equal makes me a bit uncomfortable because anything can go there since it is as uninitialized region of memory.

That said, I'm not sure what the best approach for this would be. I feel like this is a case where it would be a good idea to check the possible inputs for RBX similarly to how we wanted to check if inputs are pointers, but I am unsure.

**Comment by ccasin** [comment]
I haven't understood the behavior of the program yet, but based on Chloe's description, I agree with her assessment.  Assuming memory below RSP is equal seems maybe scary (though also possibly not the worst thing in the world if allows us to make quick progress).  Better would be to say RBX points into a region we _do_ know is equal (or are already assuming to be equal).

In general this is the kind of requirement we'd expect to have on a function (something like, "It behaves correctly IF the pointer it is passed points to an initialized, 'live' region of memory").  I certainly expect this is the kind of thing we will want to add as we develop more complex function summaries, and we'll want to check both that the function obeys this spec _and_ that when the function is called its argument satisfies these requirements.

**Comment by fortunac** [comment]
I think Chris's ideas are really good! For this specific example, we can have a precondition that RBX is pointing to a live region of memory. I guess the question would be how to express this property.

**Comment by DieracDelta** [comment]
Following our conversation yesterday, I revisited some examples. Here's a couple symbols in the grammatech examples where I think the differing pushed return address allows CBAT to choose some inputs pointing to places on the stack that result in SAT:

- `mb_copy`: This example is subtle. Differing `RBP` registers is the refuted goal. This is because the stack gets smashed using the return address. Here's a brief summary of what happens:

```
00001990: RBX := RSI
...
00001a15: mem := mem with [RSP, el]:u64 <- 0x41B7C (DIFFERING BTWN BINARIES)
00001a18: call %00128c6a with return %00001a1a
...
00001a34: RAX := pad:64[pad:32[mem[RBX + 0x10]]]
...
00001a61: mem := mem with [RBP + 0x10] <- low:8[RAX]
...
00001e78: RBP := mem[RSP, el]:u64

``` 

Note that this example is also run with `mem_offset` set to false. [Here](https://gist.github.com/96d4fc0a18b0ef7219fda247e8523e6e)'s the BIR and the returned model. Essentially `RSI` points onto the stack, and then gets read into `RBX`. This is then used  to access the return address of a function. This is then written back into memory where `RBP` was saved, resulting in `RBP` having part of the return address as its value. This is evidenced in the refuted goal:
```
Z3 Expression: = #x0000041b7c100002 #x0000041cec100002
```
- `is_dumpdir`. This is pretty representative of the bug, I think. [Here](https://gist.github.com/4e91bc8dd9cddb5e159fbd40c43dae51)'s the BIR
brief summary: 
```
000ccd85: RBX := RDI
000ccd91: mem := mem with [RSP, el]:u64 <- 0x1D749 (DIFFERING BTWN BINARIES)
000ccd94: call %00128763 with return %000ccd96
...
000ccd9b: RAX := pad:64[pad:32[mem[RBX + 0x188]]]
```
Basically, the input `RDI` points onto the stack (see the countermodel) and more specifically, is the address on the stack that a return address is pushed. This return address (that differs between programs) is then read into RAX. (Note that this countermodel was generated without mem_offset as well)
